traits/observers/_trait_event_notifier.py

```diff
@@ -0,0 +1,195 @@
+        # Reference count to avoid adding multiple equivalent notifiers
+        # to the same observable.
+        self._ref_count = 0
+
+    def __call__(self, *args, **kwargs):
+        """ Called by observables. The call signature will vary and will be
+        handled by the event factory.
+        """
+        if self.target() is None:
+            # target is deleted. The notifier is disabled.
+            return
+
+        # Hold onto the reference while invoking the handler
+        handler = self.handler()
+
+        if handler is None:
+            # The instance method is deleted. The notifier is disabled.
+            return
+
+        event = self.event_factory(*args, **kwargs)
+        if self.prevent_event(event):
+            return
+        try:
+            self.dispatcher(handler, event=event)
+        except Exception:
+            handle_exception(event)
+
+    def add_to(self, observable):
+        """ Add this notifier to an observable object.
+
+        If an equivalent notifier exists, the existing notifier's reference
+        count is bumped. Hence this method is not idempotent.
+        N number of calls to this ``add_to`` must be matched by N calls to the
+        ``remove_from`` method in order to completely remove a notifier from
+        an observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+            An object for adding this notifier to.
+        """
+        notifiers = observable._notifiers(True)
+        for other in notifiers:
+            if self.equals(other):
+                other._ref_count += 1
+                break
+        else:
+            # It is not a current use case to share a notifier with multiple
+            # observables. Using a single reference count will tie the lifetime
+            # of the notifier to multiple objects.
+            if self._ref_count != 0:
+                raise RuntimeError(
+                    "Sharing notifiers across observables is unexpected."
+                )
+            notifiers.append(self)
+            self._ref_count += 1
+
+    def remove_from(self, observable):
+        """ Remove this notifier from an observable object.
+
+        If an equivalent notifier exists, the existing notifier's reference
+        count is decremented and the notifier is only removed if
+        the count is reduced to zero.
+
+        Parameters
+        ----------
+        observable : IObservable
+            An object for removing this notifier from.
+        """
+        notifiers = observable._notifiers(True)
+        for other in notifiers[:]:
+            if self.equals(other):
+                if other._ref_count == 1:
+                    notifiers.remove(other)
+                other._ref_count -= 1
+
+                if other._ref_count < 0:
+                    raise ValueError(
+                        "Reference count is negative. "
+                        "Race condition?"
+                    )
```

Comment:
FTR, I was able to run into some kind of race conditions *sometimes* with the following test, but this particular exception has not yet occurred.
```

    def test_race_condition(self):
        dummy = DummyObservable()
        dummy.notifiers = [str] * 100
        notifier = create_notifier()

        def func1(num):
            for _ in range(num // 4):
                notifier.add_to(dummy)
            for _ in range(num // 4):
                notifier.remove_from(dummy)
            for _ in range(num // 4):
                notifier.add_to(dummy)
            for _ in range(num // 4):
                notifier.remove_from(dummy)

        import concurrent.futures
        with concurrent.futures.ThreadPoolExecutor(max_workers=60) as executor:
            list(executor.map(func1, range(10, 100)))

        self.assertEqual(len(dummy.notifiers), 100)
        self.assertEqual(notifier._ref_count, 0)
```

The last two assertions fail sometimes, the "Notifier not found." exception appears sometimes, and the "Sharing notifiers across observables is unexpected." error in `add_to` appears sometimes too.